test/test-PreviewProvider-MetadataStore.js

```diff
@@ -0,0 +1,143 @@
+/* globals require, exports */
+
+"use strict";
+
+const {before, after, waitUntil} = require("sdk/test/utils");
+const simplePrefs = require("sdk/simple-prefs");
+const {Loader} = require("sdk/test/loader");
+const loader = Loader(module);
+const httpd = loader.require("./lib/httpd");
+const {PreviewProvider} = require("lib/PreviewProvider");
+const {MetadataStore} = require("lib/MetadataStore");
+const {metadataFixture} = require("./lib/MetastoreFixture.js");
+
+const gMetadataStore = new MetadataStore();
+const gPort = 8079;
+let gPreviewProvider;
+let gPrefEmbedly = simplePrefs.prefs["embedly.endpoint"];
+let gPrefEnabled = simplePrefs.prefs["previews.enabled"];
+
+exports.test_metadatastore_saves_new_links = function*(assert) {
+  // to start, put some links in the database
+  yield gMetadataStore.asyncInsert(metadataFixture);
+  let items = yield gMetadataStore.asyncExecuteQuery("SELECT * FROM page_metadata");
+  assert.equal(items.length, 3, "sanity check that we only have 3 items in the database to start");
+
+  // these are links we are going to request - the first two should get filtered out
+  const links = [
+    {cache_key: "https://www.mozilla.org/", places_url: "https://www.mozilla.org/"},
+    {cache_key: "https://www.mozilla.org/en-US/firefox/new/", places_url: "https://www.mozilla.org/en-US/firefox/new"},
+    {cache_key: "https://notinDB.com/", places_url: "https://www.notinDB.com/", sanitized_url: "https://www.notinDB.com/"}];
+  const fakeResponse = {"urls": {
+      "https://www.notinDB.com/": {
+        "embedlyMetaData": "some embedly metadata"
+      }
+    }};
+
+  let srv = httpd.startServerAsync(gPort);
+  srv.registerPathHandler("/previewProviderMetadataStore", function handle(request, response) {
+    response.setHeader("Content-Type", "application/json", false);
+    response.write(JSON.stringify(fakeResponse));
+  });
+
+  // only request the items that are not in the database and wait for them to
+  // successfully be inserted
+  yield gPreviewProvider._asyncSaveLinks(links);
+  yield waitUntil(function*() {
+    if (gMetadataStore.transactionInProgress) {
```

Comment:
```
yield waitUntil(() => {
  return !gMetadataStore.transactionInProgress;
});
```
I wonder if it's necessary to do the waiting here, did you have any database error here? We did this in other tests, cause the Sqlite library had some buggy transaction support.